### ISSUE TITLE
[New rule] prefer-extracting-callbacks

### ISSUE THREAD

**Issue by incendial** [dart-code-checker/dart-code-metrics#360]
**Please describe what the rule should do:**
The rule should warn about inline callbacks in a widget tree and suggest to extract them to a widget methods in order to make a build method more readable. In addition extracting can help test those methods separately as well.

The rule will not trigger for lambdas, like `onPressed: () => _handler(...)`, in order to cover cases when a callback needs a variable from the outside.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[X] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**

Bad:
```dart
class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return TextButton(
      style: ...,
      onPressed: () { // LINT
        // Some 
        // Huge
        // Callback
      },
      child: ...
    );
  }
}
```

Good:
```dart
class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return TextButton(
      style: ...,
      onPressed: () => handlePressed(context),
      child: ...
    );
  }
  
  void handlePressed(BuildContext context) {
    ...
  }
}

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return TextButton(
      style: ...,
      onPressed: handlePressed,
      child: ...
    );
  }
  
  void handlePressed() {
    ...
  }
}
```

**Are you willing to submit a pull request to implement this rule?**

**Comment by incendial** [comment]
@GroovinChip please take a look to ensure that I've got it all correct.

@dkrutskikh @roman-petrov I'd really happy to hear your thoughts about this, if you have anything to add!

**Comment by GroovinChip** [comment]
Looks good to me!

**Comment by roman-petrov** [comment]
I also think it would be a great rule! 

Just one minor thing: maybe we will need some rule settings to allow "small 2-5 line length callbacks" (disabled by default), but I think we can add it later if we need it. Actually, I personally prefer not to have logic in the widget tree at all.

**Comment by incendial** [comment]
> but I think we can add it later if we need it.

Totally agreed. We should add it if only necessary.

**Comment by incendial** [comment]
@GroovinChip what do you think about `builder:` arguments, should the rule trigger on them too?

**Comment by GroovinChip** [comment]
Builder arguments can be left as is I think, because they usually build Widgets. The preference there is to build an extracted widget rather than having a big tree in the builder.

**Comment by incendial** [comment]
Okay, thanks! Then let's have `builders` excluded by default and `ignored-arguments` config parameter for custom cases.
